### PR TITLE
Hoist data pointer loads from array allocations

### PR DESCRIPTION
**Pull request by pchintalapudi**
When arrays are indexed into, the data pointer is loaded at the indexing site. Typically GVN moves this data pointer load out of loops, but we can do this more eagerly for 2D/3D array allocations and 1D array allocations that do not escape, enabling more of the optimization pipeline to come into play (e.g. loop vectorization).

This PR depends on #43487 for the initial array identification routine as well as the ArrayOpt pass, and therefore #43057 for the escape analysis improvements.

Example:
```
function f(d1, d2, d3)
           s = 0
           a = zeros(Int, d1, d2, d3)
           for i in eachindex(a)
               s += a[i]
           end
           s
       end
```

#43487 :
Godbolt: https://godbolt.org/z/9sr36nsvq
Benchmark:
```
julia> @benchmark f(100,100,100)
BenchmarkTools.Trial: 3347 samples with 1 evaluation.
 Range (min … max):  975.691 μs …   7.981 ms  ┊ GC (min … max):  0.00% …  0.00%
 Time  (median):       1.015 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.488 ms ± 836.460 μs  ┊ GC (mean ± σ):  10.71% ± 16.41%

  █▆▃                                 ▂▃▁         ▃▃         ▃▃ ▁
  ███▆▃█▄▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▄▁▁▃▁▁▁▁▃███▅▃▆▇▃▁▁▁▇██▆▅▃▇▇▃▁▁▄██ █
  976 μs        Histogram: log(frequency) by time       3.33 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 2.
```

PR:
Godbolt: https://godbolt.org/z/8qr8rYv14
Benchmark:
```
julia> @benchmark f(100,100,100)
BenchmarkTools.Trial: 4292 samples with 1 evaluation.
 Range (min … max):  688.914 μs …   6.848 ms  ┊ GC (min … max):  0.00% … 63.87%
 Time  (median):     723.949 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.159 ms ± 764.938 μs  ┊ GC (mean ± σ):  10.78% ± 17.55%

  █▆▄   ▁                          ▃▂                ▃▂    ▃▃▁  ▁
  ███▆▄▆█▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▄▄▁▄████▅▃▇█▇▄▃▁▄▁▁▃▁▃▇████▅▆████ █
  689 μs        Histogram: log(frequency) by time       2.86 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 2.
```

### CODE REVIEW

**Comment by pchintalapudi** [comment]
Since my branches are part of a fork, I've created a PR comparing this one and the array length propagation PR immediately before it here:  https://github.com/pchintalapudi/julia/pull/3

File: src/ccall.cpp [hunk 1]
```diff
@@ -1743,6 +1743,33 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
...
+            || is_libjulia_func(jl_alloc_array_3d)) {
+        jl_cgval_t retval = sig.emit_a_ccall(
+                ctx,
+                symarg,
+                argv,
+                gc_uses,
+                static_rt);
+        if (auto array_alloc = dyn_cast<CallInst>(retval.V)) {
+            array_alloc->addAttribute(AttributeList::ReturnIndex, Attribute::NoAlias);
+            array_alloc->setMetadata("allocation.array", MDNode::get(jl_LLVMContext, {}));
```
Comment (vchuravy) [inline src/ccall.cpp:1756]: I would prefer this MD to have a `julia` prefix.

File: src/llvm-alloc-helpers.cpp [hunk 2]
```diff
@@ -301,3 +301,41 @@ void jl_alloc::runEscapeAnalysis(llvm::Instruction *I, EscapeAnalysisRequiredArg
         }
     }
 }
+
+//FIXME: This doesn't actually work on Windows, as Windows inserts a trampoline
```
Comment (vchuravy) [inline src/llvm-alloc-helpers.cpp:305]: Comment is outdated now? Should instead explain why `metadata` is used

File: src/llvm-array-opt.cpp [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+Pass *createArrayOptPass()
+{
+    return new ArrayOpt();
+}
+
+extern "C" JL_DLLEXPORT void LLVMExtraAddArrayOptPass_impl(LLVMPassManagerRef PM)
+{
+    unwrap(PM)->add(createArrayOptPass());
+}
```
Comment (vchuravy) [inline src/llvm-array-opt.cpp:344]: New line missing

File: src/llvm-array-opt.cpp [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        bool elideArrayLoads();
+        bool optimizeCmpInsts();
+
+        void reset() {
+            array_allocations.clear();
+            lengths.clear();
+        }
+    };
+
+    struct ArrayOpt : public FunctionPass, public JuliaPassContext {
```
Comment (vchuravy) [inline src/llvm-array-opt.cpp:57]: Can we make this already NewPM compatible?

File: src/llvm-array-opt.cpp [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                    if (arraylen_clobbered) {
+                        break;
+                    }
+                }
+                if (!arraylen_clobbered) {
+                    //The size of the array doesn't change, we can change
+                    //length/dim loads directly to the allocated amount
+                    for (auto &field : use_info.memops) {
+                        for (auto &memop : field.second.accesses) {
+                            if (memop.offset == offsetof(jl_array_t, length) || memop.offset == offsetof(jl_array_t, nrows) || memop.offset == offsetof(jl_array_t, maxsize)) {
```
Comment (vchuravy) [inline src/llvm-array-opt.cpp:130]: Formatting, I believe convention is multiline condition if we exceed 96 chars?

**Review by vchuravy — COMMENTED** [review]
Can you add some unit-test specifically for the new llvm-array-opt pass?

**Pull request closed without merging**